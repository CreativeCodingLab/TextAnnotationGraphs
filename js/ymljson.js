// non-comprehensive function to convert yml file to json
const ymlToJson = (function() {
  function convert(url, callback) {
    var xhr = new XMLHttpRequest();
    xhr.onreadystatechange = function() {
      if (xhr.readyState === 4 && xhr.status === 200) {
        let taxonomy = [];
        let arr = taxonomy;
        const text = xhr.responseText;
        const lines = text.split('\n');
        let depths = [];

        lines.forEach(line => {
          let comment = line.indexOf('#');
          let lineStart = line.indexOf('-');
          if (lineStart < 0 || (comment >= 0 && comment < lineStart)) {
            return;
          }

          line = ((comment >= 0) ? line.slice(lineStart + 1, comment) : line.slice(lineStart + 1)).trim();

          while (depths.length > 0 && depths[depths.length - 1].i >= lineStart) {
            arr = depths.pop().arr;
          }
          if (depths.length === 0) { arr = taxonomy; }

          if (line.endsWith(':')) {
            depths.push({ arr, i: lineStart });

            let newArray = [];
            let child = {};
            child[line.slice(0, -1).trim()] = newArray;
            arr.push(child);
            arr = newArray;
          }
          else {
            arr.push(line);
          }
        });//end forEach
        if (callback) {
          callback(taxonomy);
        }
      }
    }
    xhr.open("GET", url);
    xhr.send();
  };

  return {
    convert
  };
})();
