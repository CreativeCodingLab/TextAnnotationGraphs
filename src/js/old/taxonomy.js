import ColorPicker from '../colorpicker.js';
import Word from '../components/word.js';

module.exports = (function() {
  let colors = [
    '#3fa1d1',
    '#ed852a',
    '#2ca02c',
    '#c34a1d',
    '#a048b3',
    '#e377c2',
    '#bcbd22',
    '#17becf',
    '#e7298a',
    '#e6ab02',
    '#7570b3',
    '#a6761d',
    '#7f7f7f'
  ];
  let div = {};
  let tagTypes = {};

  function updateColor(word, color) {
    if (word instanceof Word) {
      word.tag.svgText.node.style.fill = color;
    }
    else {
      word.svgText.node.style.fill = color;
    }
  };

  function updateTagColor(tag, color) {
    tagTypes[tag].forEach(word => updateColor(word, color));
  };

  class Taxonomy {
    constructor(id) {
      this.tree = {};
      div = document.getElementById('taxonomy');
    }

    draw(taxonomy, words) {
      if (taxonomy) {
        this.buildTree(taxonomy);

        if (words) {
          this.buildTagTypes(words);
        }

        this.populateTaxonomy();
        this.attachHandlers();
      }
    }

    buildTagTypes(words) {
      tagTypes = {};
      words.forEach(word => {
        if (word.tag) {
          if (tagTypes[word.tag.val]) {
            tagTypes[word.tag.val].push(word);
          }
          else {
            tagTypes[word.tag.val] = [word];
          }
        }
        if (word.clusters.length > 0) {
          word.clusters.forEach(cluster => {
            if (tagTypes[cluster.val]) {
              tagTypes[cluster.val].push(cluster);
            }
            else {
              tagTypes[cluster.val] = [cluster];
            }
          });
        }
      });
    }

    buildTree(taxonomy) {
      // turn taxonomy into a proper tree
      let flat = [];

      function createLinks(val, i, n, parent) {
        let index = { i, n };
        let obj = {
          val,
          parent,
          index: parent ? parent.index.concat(index) : [index],
          depth: parent ? parent.depth + 1 : 0,
          ancestor: parent ? parent.ancestor : null,
          children: null
        };
        if (!obj.ancestor) {
          obj.ancestor = obj;
          obj.descendantCount = 0;
        }
        ++obj.ancestor.descendantCount;

        flat.push(obj);

        if (!(typeof val === 'string' || val instanceof String)) {
          let key = Object.keys(val)[0];
          obj.val = key;
          obj.children = val[key].map((v, i) => createLinks(v, i, val[key].length, obj));
        }
        return obj;
      }

      let hierarchy = taxonomy.map((val, i) => createLinks(val, i, taxonomy.length, null));

      this.tree = {
        hierarchy,
        flat
      };
    }

    // populate modal window with list of taxonomic classes
    populateTaxonomy() {
      div.innerHTML = '<span id="toggle-taxonomy">Filter unused labels</span>';

      // build list of inputs in DOM
      let ul = document.createElement('ul');
      div.appendChild(ul);

      let nli = 1;  // number of list items
      function createLi(node, parent) {
        let li = document.createElement('li');

        // create checkbox
        let cbox = document.createElement('input');
        cbox.setAttribute('type', 'checkbox');
        cbox.id = 'cb-' + nli;

        // create label for checkbox
        let label = document.createElement('label');
        label.setAttribute('for', cbox.id);
        label.textContent = node.val;
        node.cbox = cbox;

        // create color picker input
        let picker = document.createElement('input');
        picker.className = 'colorpicker';
        picker.value = '#000000';
        picker.setAttribute('disabled', true);
        picker.node = node;
        node.picker = picker;

        li.appendChild(cbox);
        li.appendChild(label);
        li.appendChild(picker);
        parent.appendChild(li);
        ++nli;

        if (node.children) {
          let childUl = document.createElement('ul');
          li.appendChild(childUl);

          node.children.forEach(child => {
            createLi(child, childUl);
          })
        }
      }

      this.tree.hierarchy.forEach(node => {
        createLi(node, ul);
      });
    }

    // bind events to data
    attachHandlers() {
      // initialize colorpicker
      this.colorpicker = new ColorPicker('colorpicker', {
        initialColor: '#000000',
        changeCallback: (input) => {
          this.setColor(input.node);
        }
      });

      const keys = Object.keys(tagTypes);
      this.tree.flat.forEach(node => {
        // disable/enable color picking on a node
        node.cbox.onclick = () => this.onCheckboxClick(node);

        // check if tag type exists in document
        if (tagTypes[node.val]) {
          this.setColor(node, colors[keys.indexOf(node.val)]);
          node.cbox.click();
        }
      });

      // update colors of existing data
      Object.keys(tagTypes).forEach((tag, i) => updateTagColor(tag, colors[i]));
    }

    /* handle event when checkbox state changes */
    onCheckboxClick(node) {
      if (node.cbox.checked) {
        // activate node
        // propagate color to all descendants
        node.picker.removeAttribute('disabled');
        this.setColor(node);
      } else {
        // deactivate node
        // undo color propagation to all descendants
        node.picker.setAttribute('disabled', true);
        if (node.parent) {
          this.setColor(node, node.parent.picker.value);
        } else {
          this.setColor(node, '#000000');
        }
      }
    }

    /* change the color of a node */
    setColor(node, color) {
      let picker = node.picker;

      // manually set color
      if (color) {
        this.colorpicker.setColor(picker, color);
      }

      if (tagTypes[node.val]) {
        updateTagColor(node.val, picker.value);
      }

      // set color of input and descendant inputs
      function inheritColor(child) {
        if (!child.cbox.checked) {
          // set color and style
          child.picker.value = picker.value;
          child.picker.style.cssText = picker.style.cssText || child.picker.style.cssText;
          if (tagTypes[child.val]) {
            updateTagColor(child.val, picker.value);
          }

          // recursively propagate value to all children
          if (child.children) {
            child.children.forEach(inheritColor);
          }
        }
      }

      if (node.children) {
        node.children.forEach(inheritColor);
      }
    }

    remove(object) {
      // FIXME
      return;
      let tag = object.val;
      let entity = object.entity;
      if (tagTypes[tag]) {
        let i = tagTypes[tag].indexOf(entity);
        if (i > -1) {
          tagTypes[tag].splice(i, 1);
          if (tagTypes[tag].length < 1) {
            delete tagTypes[tag];
          }
        }
      }
    }

    getColor(label, object) {
      //FIXME
      return;
      let keys = Object.keys(tagTypes);
      if (tagTypes[label]) {
        return colors[keys.indexOf(label)];
      }
      else {
        tagTypes[label] = object;
        return colors[keys.length] || 'black';
      }
    }
  }
  return Taxonomy;
})();
