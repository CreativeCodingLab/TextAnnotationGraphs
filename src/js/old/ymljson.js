import load from "./xhr.js";

// non-comprehensive function to convert yml file to json
export function convert(url, callback) {
  load(url).then(function (text) {
    let taxonomy = [];
    let arr = taxonomy;
    const lines = text.split("\n");
    let depths = [];

    lines.forEach(line => {
      let comment = line.indexOf("#");
      let lineStart = line.indexOf("-");
      if (lineStart < 0 || (comment >= 0 && comment < lineStart)) {
        return;
      }

      line = ((comment >= 0) ? line.slice(lineStart + 1, comment) : line.slice(lineStart + 1)).trim();

      while (depths.length > 0 && depths[depths.length - 1].i >= lineStart) {
        arr = depths.pop().arr;
      }
      if (depths.length === 0) {
        arr = taxonomy;
      }

      if (line.endsWith(":")) {
        depths.push({arr, i: lineStart});

        let newArray = [];
        let child = {};
        child[line.slice(0, -1).trim()] = newArray;
        arr.push(child);
        arr = newArray;
      }
      else {
        arr.push(line);
      }
    });//end forEach
    if (callback) {
      callback(taxonomy);
    }
  });
}

/**
 * Converts YAML data into a JS object, taking a YAML string directly
 * instead of a URL
 * @param data
 * @return {Array}
 */
export function convertData(data) {
  let taxonomy = [];
  let arr = taxonomy;
  const lines = data.split("\n");
  let depths = [];

  lines.forEach(line => {
    let comment = line.indexOf("#");
    let lineStart = line.indexOf("-");
    if (lineStart < 0 || (comment >= 0 && comment < lineStart)) {
      return;
    }

    line = ((comment >= 0) ? line.slice(lineStart + 1, comment) : line.slice(lineStart + 1)).trim();

    while (depths.length > 0 && depths[depths.length - 1].i >= lineStart) {
      arr = depths.pop().arr;
    }
    if (depths.length === 0) {
      arr = taxonomy;
    }

    if (line.endsWith(":")) {
      depths.push({arr, i: lineStart});

      let newArray = [];
      let child = {};
      child[line.slice(0, -1).trim()] = newArray;
      arr.push(child);
      arr = newArray;
    }
    else {
      arr.push(line);
    }
  });//end forEach

  return taxonomy;
}