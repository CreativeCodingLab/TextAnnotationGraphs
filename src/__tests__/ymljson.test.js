/**
 * Test YAML parser
 */

import * as fs from "fs";

// Native implementation
import * as ymljson from "../js/old/ymljson";

// JS-YAML library
import * as yaml from "js-yaml";

describe("The `js-yaml` parser", () => {
  it("generates the same JS object as the previous implementation", () => {
    // Pull the sample taxonomy file
    const yamlData = fs.readFileSync("demo/taxonomy.yml", "utf8");

    // Native implementation
    let converted = ymljson.convertData(yamlData);
    expect(JSON.stringify(converted)).toMatchSnapshot();

    // JS-YAML library
    const converted2 = yaml.safeLoad(yamlData);
    expect(converted2).toEqual(converted);
  });
});