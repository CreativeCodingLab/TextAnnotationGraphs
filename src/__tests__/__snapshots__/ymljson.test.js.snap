// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The \`js-yaml\` parser generates the same JS object as the previous implementation 1`] = `"[\\"Alias\\",\\"ModificationTrigger\\",\\"Site\\",{\\"Context\\":[\\"Species\\",\\"CellLine\\",\\"Organ\\",\\"CellType\\",\\"Cellular_component\\",\\"TissueType\\",\\"ContextDirection\\",\\"ContextLocation\\",\\"ContextPossessive\\"]},{\\"Modification\\":[\\"PTM\\",{\\"Mutant\\":[\\"GenericMutant\\",\\"SubstitutionMutant\\",\\"DeletionMutant\\",\\"DuplicationMutant\\",\\"InsertionMutant\\",\\"FrameshiftMutant\\"]},\\"EventSite\\",\\"Negation\\"]},{\\"PossibleController\\":[{\\"Event\\":[{\\"SimpleEvent\\":[\\"Binding\\",\\"Generic_event\\",\\"Translocation\\",{\\"Amount\\":[{\\"IncreaseAmount\\":[\\"Transcription\\"]},\\"DecreaseAmount\\"]},{\\"AdditionEvent\\":[\\"Acetylation\\",\\"Farnesylation\\",\\"Glycosylation\\",\\"Hydrolysis\\",\\"Hydroxylation\\",\\"Methylation\\",{\\"Phosphorylation\\":[\\"AutoPhosphorylation\\"]},\\"Ribosylation\\",\\"Sumoylation\\",\\"Ubiquitination\\"]},{\\"RemovalEvent\\":[\\"Deacetylation\\",\\"Defarnesylation\\",\\"Deglycosylation\\",\\"Dehydrolysis\\",\\"Dehydroxylation\\",\\"Demethylation\\",\\"Dephosphorylation\\",\\"Deribosylation\\",\\"Desumoylation\\",\\"Deubiquitination\\"]}]},{\\"ComplexEvent\\":[{\\"Regulation\\":[\\"Positive_regulation\\",\\"Negative_regulation\\"]},{\\"ActivationEvent\\":[\\"Positive_activation\\",\\"Negative_activation\\"]}]}]},{\\"Entity\\":[{\\"BioEntity\\":[\\"BioProcess\\",{\\"BioChemicalEntity\\":[\\"Generic_entity\\",\\"Simple_chemical\\",{\\"Equivalable\\":[\\"Family\\",{\\"MacroMolecule\\":[\\"Protein\\",\\"Gene_or_gene_product\\",\\"Complex\\",\\"GENE\\"]}]}]}]}]}]},{\\"Precedence\\":[\\"DirectConnection\\",\\"IndirectConnection\\",{\\"BetweenSentence\\":[\\"TimexAfter\\",\\"TimexBefore\\",\\"SentenceInitialEvent\\",\\"InterAfter\\",\\"InterBefore\\"]},{\\"TAM\\":[\\"Aux\\",{\\"Tense\\":[\\"PastTense\\",\\"PresentTense\\",\\"FutureTense\\"]},{\\"Aspect\\":[\\"Perfective\\",\\"Progressive\\"]}]}]}]"`;
